auth/src/main/java/com/firebase/ui/auth/data/remote/FacebookSignInHandler.java

```diff
@@ -0,0 +1,167 @@
+
+        // Ensure we have email and public_profile scopes
+        if (!permissionsList.contains(EMAIL)) {
+            permissionsList.add(EMAIL);
+        }
+        if (!permissionsList.contains(PUBLIC_PROFILE)) {
+            permissionsList.add(PUBLIC_PROFILE);
+        }
+
+        mPermissions = permissionsList;
+    }
+
+    public List<String> getPermissions() {
+        return mPermissions;
+    }
+
+    @Override
+    public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
+        mCallbackManager.onActivityResult(requestCode, resultCode, data);
+    }
+
+    @Override
+    public void onSuccess(final LoginResult result) {
+        GraphRequest request = GraphRequest.newMeRequest(result.getAccessToken(),
+                new ProfileRequest(result));
+
+        Bundle parameters = new Bundle();
+        parameters.putString("fields", "id,name,email,picture");
+        request.setParameters(parameters);
+        request.executeAsync();
+    }
+
+    @Override
+    public void onCancel() {
+        onError(new FacebookException());
+    }
+
+    @Override
+    public void onError(FacebookException e) {
+        setResult(IdpResponse.fromError(e));
+    }
+
+    @Override
+    protected void onCleared() {
+        super.onCleared();
+        LoginManager.getInstance().unregisterCallback(mCallbackManager);
+    }
+
+    private class ProfileRequest implements GraphRequest.GraphJSONObjectCallback {
+        private final LoginResult mResult;
+
+        public ProfileRequest(LoginResult result) {
+            mResult = result;
+        }
+
+        @Override
+        public void onCompleted(JSONObject object, GraphResponse response) {
+            FacebookRequestError error = response.getError();
+            if (error != null || object == null) {
+                setResult(IdpResponse.fromError(error == null ? new FirebaseUiException(
+                        ErrorCodes.PROVIDER_ERROR, "Facebook graph request failed")
+                        : error.getException()));
+                return;
+            }
+
+            String email = null;
+            String name = null;
+            Uri photoUri = null;
+
+            try {
+                email = object.getString("email");
+            } catch (JSONException ignored) {}
+            try {
+                name = object.getString("name");
+            } catch (JSONException ignored) {}
+            try {
+                photoUri = Uri.parse(object.getJSONObject("picture")
+                        .getJSONObject("data")
+                        .getString("url"));
+            } catch (JSONException ignored) {}
+
+            setResult(createIdpResponse(mResult, email, name, photoUri));
+        }
+    }
+
+    public static final class Params extends ProviderHandlerParamsBase {
+        private final AuthUI.IdpConfig config;
+
+        public Params(ProvidersHandlerBase handler, AuthUI.IdpConfig config) {
```

Comment:
Strange to see a static inner class taking an instance of the parent class ... what's going on here?